### PR TITLE
feat: extract metrics from mlflow 

### PR DESCRIPTION
**Pull request by HLasse**
Wanted to create a cute little table of performances from my text experiment, but turned out to be more painful than expected to interface with mlflow. Created this little class to help. 

Very open for comments, especially on naming and placement in the repo

### CODE REVIEW

File: psycop/projects/scz_bp/text_experiment/mlflow_data_extraction.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            pl.lit(run.info.run_name).alias("run_name")
+        )
+
+    def _get_mlflow_experiment_id_from_experiment_name(self, experiment_name: str) -> str:
+        experiment = self.client.get_experiment_by_name(name=experiment_name)
+        if experiment is None:
+            raise ValueError(f"{experiment_name} does not exist on MlFlow.")
+        return experiment.experiment_id
+
+    def _get_mlflow_runs_by_experiment(self, experiment_name: str) -> PagedList[Run]:
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/text_experiment/mlflow_data_extraction.py:55]: What is the consequence of a PagedList here?

File: psycop/projects/scz_bp/text_experiment/mlflow_data_extraction.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+from collections.abc import Iterable
+
+import polars as pl
+from mlflow.entities import Run
+from mlflow.store.entities import PagedList
+from mlflow.tracking import MlflowClient
+
+
+class MlflowDataExtraction:
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/text_experiment/mlflow_data_extraction.py:9]: `MLFlowMetricExtractor`?

**Comment by MartinBernstorff** [comment]
Looks great! Seems like an annoying problem, with a nice solution.

I'm likely biased, but would love the returntypes to be a subclass of `ValidatedFrame`! 

Then we can read the dataframe specification from the types instead of a docstring, and we ensure it stays up-to-date 👍 

What do you think here?

**Comment by MartinBernstorff** [comment]
We could also test this by mocking the client, but I personally don't find it that important here 👍

**Comment by HLasse** [comment]
Yes, strong agree on the ValidatedFrame - I thought the same, but wanted your thoughts first!

File: psycop/projects/scz_bp/text_experiment/mlflow_data_extraction.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            self._get_metric_for_run(run_id=run.info.run_id, metric=metric) for metric in metrics
+        ]
+        return pl.concat(metrics_df).with_columns(pl.lit(run.info.run_name).alias("run_name"))
+
+    def _get_metric_for_run(self, run_id: str, metric: str) -> pl.DataFrame:
+        return pl.DataFrame(
+            [
+                dict(metric_object)
+                for metric_object in self.client.get_metric_history(run_id=run_id, key=metric)
+            ]
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/text_experiment/mlflow_data_extraction.py:41]: Just want to highlight an Iterpy solution here:

`return Iter(self.client.get_metric_history(run_id=run_id, key=metric)).map(dict).to_list()`

Comment (MartinBernstorff) [reply psycop/projects/scz_bp/text_experiment/mlflow_data_extraction.py:41]: Huh, I expected that to be much more readable, but it was only OK 🤷‍♂️

File: psycop/projects/scz_bp/text_experiment/mlflow_data_extraction.py (hunk 1 above)
Comment (HLasse) [reply psycop/projects/scz_bp/text_experiment/mlflow_data_extraction.py]: That's the return type of the search function. Doesn't matter for us - changed to iterable

File: psycop/projects/scz_bp/text_experiment/mlflow_data_extraction.py (hunk 3 above)
Comment (HLasse) [reply psycop/projects/scz_bp/text_experiment/mlflow_data_extraction.py]: Hm, yeah - not sure which I prefer tbh 🤷‍♂️

**Comment by HLasse** [comment]
Also added functionality to retrieve the best run across one or more experiments 👍

**Pull request merged**